File: backend/FFXIVLoot.Application/Services/WeekService.cs
using FFXIVLoot.Application.DTOs;
using FFXIVLoot.Application.Interfaces;
using FFXIVLoot.Domain.Interfaces;

namespace FFXIVLoot.Application.Services;

/// <summary>
/// Service for managing raid weeks
/// </summary>
public class WeekService : IWeekService
{
    private readonly IWeekRepository _weekRepository;

    /// <summary>
    /// Initializes a new instance of WeekService
    /// </summary>
    public WeekService(IWeekRepository weekRepository)
    {
        _weekRepository = weekRepository ?? throw new ArgumentNullException(nameof(weekRepository));
    }

    /// <summary>
    /// Gets the current active week
    /// </summary>
    public async Task<WeekDto?> GetCurrentWeekAsync()
    {
        var week = await _weekRepository.GetCurrentWeekAsync();
        if (week == null)
        {
            return null;
        }

        return new WeekDto
        {
            WeekNumber = week.WeekNumber,
            StartedAt = week.StartedAt,
            IsCurrent = week.IsCurrent
        };
    }

    /// <summary>
    /// Gets all weeks
    /// </summary>
    public async Task<List<WeekDto>> GetAllWeeksAsync()
    {
        var weeks = await _weekRepository.GetAllAsync();
        return weeks.Select(w => new WeekDto
        {
            WeekNumber = w.WeekNumber,
            StartedAt = w.StartedAt,
            IsCurrent = w.IsCurrent
        }).OrderByDescending(w => w.WeekNumber).ToList();
    }

    /// <summary>
    /// Starts a new week
    /// </summary>
    public async Task<WeekDto> StartNewWeekAsync()
    {
        var allWeeks = await _weekRepository.GetAllAsync();
        var maxWeekNumber = allWeeks.Any() ? allWeeks.Max(w => w.WeekNumber) : 0;
        var newWeekNumber = maxWeekNumber + 1;

        await _weekRepository.CreateWeekAsync(newWeekNumber);
        await _weekRepository.SetCurrentWeekAsync(newWeekNumber);

        var week = await _weekRepository.GetCurrentWeekAsync();
        if (week == null)
        {
            throw new InvalidOperationException("Failed to create new week");
        }

        return new WeekDto
        {
            WeekNumber = week.WeekNumber,
            StartedAt = week.StartedAt,
            IsCurrent = week.IsCurrent
        };
    }

    /// <summary>
    /// Sets a specific week as the current week
    /// </summary>
    public async Task SetCurrentWeekAsync(int weekNumber)
    {
        await _weekRepository.SetCurrentWeekAsync(weekNumber);
    }

    /// <summary>
    /// Creates a week with a specific week number
    /// </summary>
    public async Task<WeekDto> CreateWeekWithNumberAsync(int weekNumber)
    {
        var week = await _weekRepository.CreateWeekAsync(weekNumber);
        
        return new WeekDto
        {
            WeekNumber = week.WeekNumber,
            StartedAt = week.StartedAt,
            IsCurrent = week.IsCurrent
        };
    }

    /// <summary>
    /// Deletes a week by week number
    /// </summary>
    public async Task DeleteWeekAsync(int weekNumber)
    {
        await _weekRepository.DeleteWeekAsync(weekNumber);
    }
}

File: frontend/src/services/api/weekService.ts
import { apiRequest } from './apiClient';
import { Week } from '../../types/member';

const WEEK_ENDPOINT = '/api/weeks';

/**
 * Service for managing raid weeks
 */
export const weekService = {
  /**
   * Gets the current active week
   */
  async getCurrentWeek(): Promise<Week> {
    return apiRequest<Week>(`${WEEK_ENDPOINT}/current`);
  },

  /**
   * Gets all weeks
   */
  async getAllWeeks(): Promise<Week[]> {
    return apiRequest<Week[]>(WEEK_ENDPOINT);
  },

  /**
   * Starts a new week
   */
  async startNewWeek(): Promise<Week> {
    return apiRequest<Week>(`${WEEK_ENDPOINT}/new`, {
      method: 'POST',
    });
  },

  /**
   * Sets a specific week as the current week
   */
  async setCurrentWeek(weekNumber: number): Promise<void> {
    return apiRequest<void>(`${WEEK_ENDPOINT}/${weekNumber}/set-current`, {
      method: 'POST',
    });
  },

  /**
   * Creates a week with a specific week number
   */
  async createWeekWithNumber(weekNumber: number): Promise<Week> {
    return apiRequest<Week>(`${WEEK_ENDPOINT}/create/${weekNumber}`, {
      method: 'POST',
    });
  },

  /**
   * Deletes a week and reverts all BiS tracker changes
   */
  async deleteWeek(weekNumber: number): Promise<void> {
    return apiRequest<void>(`${WEEK_ENDPOINT}/${weekNumber}`, {
      method: 'DELETE',
    });
  },
};

